ROSnode for line detection

camkisailus opened the pull request.

camkisailus committed:
Updated line_detector.py to work with realsense stream

camkisailus committed:
Lint

camkisailus committed:
Lint 2x

Sridhar701Pitt committed:
ROS implementation, added TODOs and relevant comments

noah710 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,142 @@
-                [right_x_start, max_y, right_x_end, min_y],
-            ]],
-            thickness=5,
+class Obstacle:
```
This class is never used. I think you'd be better off making another script for obstacles.

nls5260 replied:
I think it would be good if we could have a separate rosnode that publishes info on obstacles.

asaba96 replied:
Agreed

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,142 @@
-# import matplotlib.pyplot as plt
+import pyrealsense2 as rs
 
 ddepth = cv2.CV_16S
```
These magic numbers should be loadable parameters

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,142 @@
-                [right_x_start, max_y, right_x_end, min_y],
-            ]],
-            thickness=5,
+class Obstacle:
```
Since we are using python 2, it is good practice to inherit from `object`

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,142 @@
+class Obstacle:
+
+    def __init__(self, type_of_obstacle, color_of_obstacle, x_center, y_center):
+        self.type_of_obstacle = type_of_obstacle
```
In python, private variables and methods are named with a `_` at the beginning. Ex. `self._type_of_obstacle`

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,142 @@
+        self.dist = -1  # default to -1 so we can check later if the distance has been verified
+
+
+class Lines:
```
Same as above, inherit from object

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,142 @@
+class Lines:
+
+    def __init__(self):
+        self.centers = 0
```
Same as above for private variables and methods. Add a `_` at the beginning. Ex. `self._type_of_obstacle`

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,142 @@
+
+    def skeletize(self, img, size, skel):
+        element = cv2.getStructuringElement(cv2.MORPH_CROSS, (3, 3))
+        done = False
```
Please fix this logic. Additionally, never use `while True` always check if `rospy.is_shutdown()`

asaba96 replied:
Also, you don't need the parentheses after the while

asaba96 replied:
Nitpicky, but this can still be cleaned up some more @camkisailus

asaba96 replied:
@camkisailus please fix

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,142 @@
+                done = True
+        return skel
+
+    def draw_lines(img, lines, color=[0, 255, 0], thickness=3):
```
Same as above for private variables and methods. Add a `_` at the beginning. Ex. `self._type_of_obstacle`

camkisailus replied:
fixed

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,142 @@
+        img = cv2.addWeighted(img, 0.8, line_img, 1.0, 0.0)
+        return img
+
+    def best_fit(x_points, y_points):
```
Same as above for private variables and methods. Add a `_` at the beginning. Ex. `self._type_of_obstacle`

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,142 @@
+        dst = cv2.Laplacian(skeleton, ddepth, ksize=kernel_size)
+        abs_dst = cv2.convertScaleAbs(dst)
+
+        linesP = cv2.HoughLinesP(abs_dst, rho=6, theta=np.pi / 60, threshold=50,
```
I see a lot of magic numbers here.  Please make them params.

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,142 @@
+                        cv2.circle(cv_image, (x2, y2), (5), (0, 255, 0), 3)
+
+        cv2.namedWindow('Detections', cv2.WINDOW_AUTOSIZE)
+        cv2.imshow('Detections', cv_image)
```
This is a good debugging script, but anything production shouldn't be showing anything. We need a ROS-ified version at some point please.

noah710 reviewed (dismissed):
This is just a python script right now. We should hold off on merging it until it runs as a node. I think we should have it publish some raw data about the detections so we can work on transforming that data to real world coordinates.

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,142 @@
+
+
+def main():
+    pipeline = rs.pipeline()
```
With the ROS version, we would be using realsense ROS instead of streaming direct from the camera. If we do it this way, only one process can use the camera images.

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,142 @@
+    config.enable_stream(rs.stream.depth, 640, 480, rs.format.z16, 30)
+    pipeline.start(config)
+    go = True
+    while go:
```
same comments as above

asaba96 reviewed (changes requested):
This is a good first start, but we need a cleaned up, ros-ified production version.

Sridhar701Pitt committed:
ROS implementation

camkisailus committed:
First step of ROS Implementation

camkisailus committed:
Commented for fixes needed

Sridhar701Pitt committed:
ROS Implementation: Subscribes from the real sense ros wrapper node and published processed image to a node

Sridhar701Pitt committed:
Updated comments and TODOs and unnecessary code deletions

Sridhar701Pitt committed:
Adding private variables and some cleanup

Sridhar701Pitt committed:
Lint Cleanup and CvBridge try catch addition

Sridhar701Pitt committed:
Modified private variables

camkisailus committed:
Added point array publisher

Sridhar701Pitt committed:
Lint + Update private variables in Lines class (Verify if working)

asaba96 commented on src/magellan_vision/init.sh:
```diff
@@ -0,0 +1,21 @@
+#!/bin/sh
+# TODO: Script file description
+###source setup file befor running this
```
Please remove this script and use the launch setup

noah710 replied:
fixed

asaba96 replied:
@noah710 this is still here whats up with that

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,159 @@
+from magellan_core.msg import Float64Arr  # ROS Float64Arr type msg
+
+# Global Constant
+VERBOSE = True
```
This should not be `True` by default. Additionally, it should be a parameter loaded by ROS so we can change it in a launch file.

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,159 @@
+        # subscribed Topic
+        self._subscriber = rospy.Subscriber("/camera/color/image_raw", Image, self._callback, queue_size=1)
+        if VERBOSE:
+            print "subscribed to /camera/color/image_raw"
```
Please use `rospy.log_debug` or log_info

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,159 @@
+        '''Callback function of subscribed topic.
+        Here images get converted and features detected and published'''
+        if VERBOSE:
+            print 'received image of type: "%s"' % type(ros_data)
```
Same as above

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,159 @@
+        image_np = self._lines_object.detect(image_np)
+        time2 = time.time()
+        if VERBOSE:
+            print 'Detection processed at %s Hz.' % (1/(time2-time1))
```
same as above

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,159 @@
+        self._point_arr_pub.publish(self._lines_object.points_arr)
+        time3 = time.time()
+        if VERBOSE:
+            print 'Subscribe to publish frequency: %s Hz' % (1/(time3-time0))
```
same as above

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,159 @@
+#                   to the /perception/color/image_processed topic(Processed RGB raw Images) and to the
+#                   /perception/detectedPoints topic(Detected Points as a Float64Arr msg)
+# TODO: Publish to other topics: Depth image; etc..
+class PubSubNode:
```
Needs to inherit from `object`

noah710 replied:
fixed

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,159 @@
+    def skeletize(self, img, size, skel):
+        element = cv2.getStructuringElement(cv2.MORPH_CROSS, (3, 3))
+        done = False
+        while(not done):
```
Still needs addressed from earlier comments.

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,159 @@
+    # disable_signals flag allows catching signals(Excecptions) such as the KeyboardInterrupt, otherwise try/except
+    # Exceptions may never be handled
+
+    rospy.init_node('LinesNode', anonymous=True, disable_signals=True)
```
Why is `anonymous=True`? We expect only one running copy of this node to be running at a time. Additionally, `disable_signals` refers to the ability to handle your own signals in the event of a shutdown. Please leave it at the default. From http://wiki.ros.org/rospy/Overview/Initialization%20and%20Shutdown 

```
By default, rospy registers signal handlers so that it can exit on Ctrl-C. In some code, you may wish to disable this, including:
You're not calling init_node() from the Python Main thread. Python only allows signals to be registered from the Main thread.
You're running rospy within Qt, wxPython or any other GUI toolkit that has its own exit handlers.
You wish to have your own signal handlers by default.
```

noah710 replied:
fixed

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,159 @@
+    try:
+        rospy.spin()
+    except KeyboardInterrupt:
+        print "shutting down ROS Lines detector Module"
```
Same as above, do not use print.

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,159 @@
+
+        # Publish Processed Image
+        self._image_pub.publish(ros_msg)
+        self._point_arr_pub.publish(self._lines_object.points_arr)
```
Where are you creating the ros msg before publishing the array? Have you tested that this works?

noah710 replied:
This seems to work now

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,159 @@
+        self._lines_object = Lines()
+        # topic where we publish
+        self._image_pub = rospy.Publisher("/perception/color/image_processed", Image, queue_size=5)
+        self._point_arr_pub = rospy.Publisher('/perception/detectedPoints', Float64Arr, queue_size=10)
```
Can you set the topic name to something like `detected_points`? It's not common to use camel case for topic names.

noah710 replied:
fixed

asaba96 commented on src/magellan_core/msg/Float64Arr.msg:
```diff
@@ -0,0 +1 @@
+float64[] data
```
So the ROS message naming convention would mean that this message should be named `Float64Array`. Additionally, we should be adding a header to signify what the frame is and when the lines were detected. I would follow what Levi did here https://github.com/Pitt-RAS/iarc7_msgs/blob/master/msg/Float64ArrayStamped.msg

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,159 @@
+        rospy.spin()
+    except KeyboardInterrupt:
+        print "shutting down ROS Lines detector Module"
+    cv2.destroyAllWindows()
```
We shouldn't be creating any windows in this node so we shouldn't be calling this.

noah710 replied:
fixed

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,159 @@
+    PubSubNode()
+    try:
+        rospy.spin()
+    except KeyboardInterrupt:
```
Change to catch-all `Exception` type

asaba96 replied:
and change log to log fatal

4ndr3w commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,159 @@
+VERBOSE = True
+
+
+# Class PubSubNode: Subscribes to the /camera/color/image_raw topic(RGB Images from realsense camera) and publishes
```
Consider putting this information in a docstring for this class: https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings

4ndr3w commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,159 @@
+        # conversion to cv2
+        try:
+            image_np = self._bridge.imgmsg_to_cv2(ros_data, "bgr8")
+        except CvBridgeError:
```
It might be worth logging if this fails

asaba96 replied:
@camkisailus how was this resolved, I don't see the log

asaba96 replied:
@camkisailus this should not be a debug message, it should be an error

Sridhar701Pitt committed:
Updated points Msg, updated CMakeLists.txt, updated line_detector.py and it works

- Message now contains X and Y points seperately in array format
- De linted and now works with the real sense camera

noah710 committed:
added launch file

noah710 committed:
inherit from object

noah710 committed:
minor fixes

camkisailus committed:
Fixed

camkisailus committed:
linter

camkisailus committed:
lint

asaba96 commented on src/magellan_core/CMakeLists.txt:
```diff
@@ -59,6 +59,7 @@ add_message_files(
   ObstacleStamped.msg
   ObstacleStampedArray.msg
   EncoderCount.msg
+  Int32XYArr.msg
```
Per discussion with @camkisailus let's please just have an array of [Points](http://docs.ros.org/melodic/api/geometry_msgs/html/msg/Point.html) for the line points.

camkisailus replied:
@Sridhar701Pitt Is claiming this

asaba96 replied:
@camkisailus @Sridhar701Pitt please fix this

asaba96 commented on src/magellan_core/launch/line_detector.launch:
```diff
@@ -0,0 +1,3 @@
+<launch>
+        <node pkg="magellan_vision" type="line_detector.py" name="line_detector_node"/>
+</launch>
```
No newline at EOF

asaba96 commented on src/magellan_core/msg/Int32XYArr.msg:
```diff
@@ -0,0 +1,2 @@
+int32[] pixelX
```
See comment above

camkisailus replied:
@Sridhar701Pitt

asaba96 replied:
@camkisailus @Sridhar701Pitt please fix

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,170 @@
+/output/color/image_processed topic(Processed RGB raw Images) and to the /perception/detetedPoints topic
+(Detected points as a Float64Arr msg)
+
+#TODO: Publish to other topics such as Depth Image.
```
This TODO and run instructions need to be removed.

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,170 @@
+from magellan_core.msg import Int32XYArr  # ROS Float64Arr type msg
+
+# Global Constant
+VERBOSE = False
```
Again, please make parameter

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,170 @@
+            rospy.logdebug('Subscribe to publish frequency: %s Hz' % (1/(time3-time0)))
+
+
+# Class Lines: Detects points in an Image that aligns into a line. Outputs the image with detected points overlayed
```
Docstring style please

camkisailus replied:
fixed I think

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,170 @@
+    # disable_signals flag allows catching signals(Excecptions) such as the KeyboardInterrupt, otherwise try/except
+    # Exceptions may never be handled
+
+    rospy.init_node('LinesNode', anonymous=False)
```
Just remove setting the `anonymous` flag, it is False by default

camkisailus committed:
params added and implemented

camkisailus committed:
lint

camkisailus committed:
minor fixes

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -28,13 +16,16 @@
 
 # Global Constant
-VERBOSE = False
+VERBOSE = rospy.get_param("VERBOSE")
```
Move these to be members of the objects here, not global constants.

asaba96 replied:
Additionally, when you do, wrap them in a `try...except KeyError` and log error that you failed to get a param and crash the node (i.e. `raise`)

noah710 committed:
quickfix

asaba96 commented:
@camkisailus @Sridhar701Pitt please rebase this branch

camkisailus committed:
Lowered CPU usage

camkisailus committed:
Linters exist for a reason

camkisailus committed:
Lint 2x

asaba96 commented on src/magellan_core/launch/line_detector.launch:
```diff
@@ -0,0 +1,10 @@
+<launch>
+        <node pkg="magellan_vision" type="line_detector.py" name="line_detector_node"/>
+        <param name="VERBOSE" type="boolean" value="false"/>
```
can we change this to lower case like everything else?

camkisailus replied:
fixed

asaba96 commented on src/magellan_core/launch/passive.launch:
```diff
@@ -2,7 +2,7 @@
 
     <node pkg="magellan_core" type="viz_turning_radius.py" name="viz_turning_radius" />
-
+  
```
why are there whitespace changes in here? can we clean this up

camkisailus replied:
fixed

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,156 @@
+        self._bridge = CvBridge()
+
+        # subscribed Topic
+        self._subscriber = rospy.Subscriber("/camera/color/image_raw", Image, self._callback, queue_size=1)
```
Not sure how I feel about specifying the `queue_size` in the subscriber. This ensures we drop any old messages and only get the latest image I guess. @4ndr3w @noah710 any thoughts?

noah710 replied:
I thought this was a good idea. We'll have to test it @cuhsailus

4ndr3w replied:
I think this is fine, we really only care about the latest

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,156 @@
+        self._lines_object = Lines()
+        # topic where we publish
+        self._image_pub = rospy.Publisher("/perception/color/image_processed", Image, queue_size=5)
+        # self._point_arr_pub = rospy.Publisher('/perception/detected_points', Int32XYArr, queue_size=5)
```
We need to publish the points, can this please be fixed? @camkisailus @Sridhar701Pitt we need a different message type, per our discussion before.

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,156 @@
+        self._image = None
+
+        if self._VERBOSE:
+            rospy.logdebug("subscribed to /camera/color/image_raw")
```
can we delete this entirely, it's not needed.

camkisailus replied:
deleted

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,156 @@
+                    return
+
+                # Publish Processed Image amnd Points
+                self._image_pub.publish(ros_msg)
```
We might one to have a param to control whether or not we publish the debug image

noah710 replied:
@cuhsailus added the verbose param

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,156 @@
+                # self._point_arr_pub.publish(intXYMsg)
+
+    '''Callback function of subscribed topic.
+    Here images get converted and features detected and published'''
```
comment needs updated.

camkisailus replied:
I'm just going to remove this comment

noah710 replied:
@asaba96 these are fixed just not pushed so I'm resolving

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,156 @@
+            try:
+                self._image = self._bridge.imgmsg_to_cv2(ros_data, "bgr8")
+            except CvBridgeError:
+                rospy.logdebug('Could not convert image')
```
This should be logerr not log debug

camkisailus replied:
fixed

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,156 @@
+    def _callback(self, ros_data):
+        with self._lock:
+            if self._VERBOSE:
+                rospy.logdebug('received image of type: "%s"' % type(ros_data))
```
can we delete this entirely, it's not needed.

camkisailus replied:
deleted

camkisailus replied:
deleted

asaba96 commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,156 @@
+def main(args):
+    rospy.init_node('LinesNode')
+    node_ = PubSubNode()
+    rate = rospy.Rate(20)
```
please make rate a param

camkisailus replied:
fixed

asaba96 commented:
Also this needs rebased

noah710 commented:
Rebased

noah710 committed:
remove init.sh

camkisailus committed:
PR comments

noah710 committed:
catch all

noah710 committed:
PointArray msg

noah710 committed:
linter no like

4ndr3w commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,170 @@
+        self._image = None
+
+    def run_detects(self):
+        with self._lock:
```
Might be worth just grabbing a reference to `self._image` here so we don't keep the lock in `_callback` blocked while we process this image. Right after you unblock `_callback`, you'll get whatever the first image passed to the callback was - probably a stale image when you get to it. 

Probably not a big deal overall, but will get you slightly fresher images.

asaba96 replied:
I agree with Lobos, we should probably do that

4ndr3w commented on src/magellan_core/launch/passive.launch:
```diff
@@ -10,6 +10,12 @@
     <include file="$(find magellan_core)/launch/navigation.launch" />
     <!--<include file="$(find magellan_core)/launch/fake_path.launch" />-->
     <!--<include file="$(find magellan_core)/launch/obstacle_avoidance.launch" />-->
+<<<<<<< HEAD
```
Probably want to clean this up

4ndr3w commented on src/magellan_vision/scripts/line_detector.py:
```diff
@@ -1,211 +1,170 @@
+            rate.sleep()
+
+    except:
+        rospy.logfatal("shutting down ROS Lines detector Module")
```
Would be nice to print the exception, since this is basically a catch-all for the entire program

camkisailus committed:
Fixed Point msg

camkisailus committed:
Merge branch 'line_detection_from_stream' of https://www.github.com/Pitt-RAS/racecar into line_detection_from_stream

camkisailus committed:
cleaning passive and adding error print to line detector

noah710 committed:
Fixed launch
